backend/substrapp/tasks/task.py

```diff
@@ -0,0 +1,98 @@
+"""
+This file contains the main logic for executing a compute task:
+
+- Create execution context
+- Populate asset buffer
+- Loads assets from the asset buffer
+- **Execute the compute task**
+- Save the models/results
+- Teardown the context
+
+We also handle the retry logic here.
+"""
+import enum
+import pickle  # nosec B403
+from typing import Any
+
+import structlog
+from billiard.einfo import ExceptionInfo
+from celery import Task
+from django.conf import settings
+
+import orchestrator
+from substrapp.compute_tasks.compute_pod import delete_compute_plan_pods
+from substrapp.models import FailedAssetKind
+from substrapp.task_routing import WORKER_QUEUE
+from substrapp.tasks.tasks_asset_failure_report import store_asset_failure_report
+
+logger = structlog.get_logger(__name__)
+
+
+class FailableTask(Task):
+    asset_type: FailedAssetKind
+
+    def on_failure(
+        self, exc: Exception, task_id: str, args: tuple, kwargs: dict[str, Any], einfo: ExceptionInfo
+    ) -> None:
+        asset_key, channel_name = self.get_task_info(args, kwargs)
+        exception_pickled = pickle.dumps(exc)
```

Comment:
I'd add a comment as to why we need to pickle the exception (for future reference)